include/dgl/immutable_graph.h

```diff
@@ -0,0 +1,480 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file dgl/graph.h
+ * \brief DGL immutable graph index class.
+ */
+#ifndef DGL_IMMUTABLE_GRAPH_H_
+#define DGL_IMMUTABLE_GRAPH_H_
+
+#include <vector>
+#include <string>
+#include <cstdint>
+#include <utility>
+#include <tuple>
+#include "runtime/ndarray.h"
+#include "graph.h"
+
+namespace dgl {
+
+struct ImmutableSubgraph;
+struct SampledSubgraph;
+
+template<class ForwardIt, class T>
+bool binary_search(ForwardIt first, ForwardIt last, const T& value) {
+  first = std::lower_bound(first, last, value);
+  return (!(first == last) && !(value < *first));
+}
+
+/*!
+ * \brief Base dgl immutable graph index class.
+ *
+ */
+class ImmutableGraph {
```

Comment:
I think we could define an interface for graph interface and ImmutableGraph simply inherits that. As a result, we don't need to:
(1) Expose all the related CAPI again.l
(2) The GraphIndex class in python could be resued for immutable graph.
Thoughts?

Reply:
I was struggling with it. There are a few things that these two indices are incompatible.
GraphIndex supports:
* graph modification,
* FindEdge (this operation is very slow in CSR, so I didn't implement it).
* something like this `const std::vector<dgl_id_t>& SuccVec` is not possible for ImmutableGraphIndex.

ImmutableGraphIndex supports:
* sampling,
* Get CSR arrays,

We could merge the CAPIs for the basic operations.

Reply:
For graph modification, we could throw errors.

For `FindEdge`, it's OK to have very naive implementation.

For `SuccVec`, we could change it. It is used to implement graph traversal (not exposed to CAPI at all). What would you suggest? Please read how it is used in graph traversal.

For `sampling` and Get CSR arrays, we could put it in the new graph class not in the base class. And when the graph pointer is passed down from python, we just cast it to the new class pointer.

Strongly suggest we define a common abstract right now. It would be very painful to do that later.

Also, change `ImmutableGraphIndex` to `CSRGraphIndex`? We might have more formats later.

Reply:
I'm in line with @jermainewang .

Also is there a particular reason why (Mutable)GraphIndex *does not* support samling?  In fact I think it makes sense to support sampling there as well, since that will be helpful for learning on large dynamic graphs in the (far) future.

Reply:
Sampling requires very efficient implementation. Mutable graph index doesn't allow efficient implementation.

Reply:
I'm fine with implementing a common interface.

Actually ImmutableGraphIndex contains both CSR and CSC for the graph.

Reply:
The main problem with `SuccVec` is that it returns a reference. ImmutableGraphIndex can't return a reference. One option is to return `std::vector` directly. however, this may cause unnecessary overhead for GraphIndex.

Reply:
How about an iterator?

Reply:
@zheng-da I assume that "efficient implementation" refers to storing the graph in CSR.  Could you point out which part of the sampling depends on CSR in particular?  I went through the code and didn't find it..

Reply:
The efficiency comes from the format. CSR stores data in a single piece of contiguous memory, which makes memory management much easier. The mutable graph index stores the adjacnecy lists in individual vectors, which results in many small memory allocation. As far as I remember, subgraph construction is significantly faster on CSR than on the mutable graph index.

Reply:
In this case, should we switch to the single contiguous memory solution for the Mutable graph index as well?  igraph goes this way, and therefore adding edges takes linear time of the number of edges of the new graph.  I guess this is okay since we don't have a lot of separate edge additions in general.

Reply:
Our original graph index was designed for general purpose. I think the current impl is quite reasonable. I think it's better to add a function that converts a general graph index into a read-only graph index if someone really wants to sample from the general graph index.